lib/ClangImporter/ImportDecl.cpp

```diff
@@ -1974,6 +1975,59 @@ applyPropertyOwnership(VarDecl *prop,
   }
 }
 
+/// Turn T into Unmanaged<T>, or into Unmanaged<T>? if `isOptional` is true.
+static Type getUnmanagedType(swift::ASTContext& Ctx,
+                             Type payloadType, bool isOptional) {
+  NominalTypeDecl *unmanagedDecl = Ctx.getUnmanagedDecl();
+  if (!unmanagedDecl || unmanagedDecl->getGenericParams()->size() != 1)
+    return payloadType;
+
+  Type result = BoundGenericType::get(unmanagedDecl, /*parent*/ Type(),
+                                      payloadType);
+  if (isOptional)
+    result = OptionalType::get(result);
+  return result;
+}
+
+static void
+applyPropertyOwnership(VarDecl *prop,
+                       clang::Qualifiers::ObjCLifetime lifetime) {
+  Type ty = prop->getInterfaceType();
+  bool wasOptional = false;
+  if (auto innerTy = ty->getOptionalObjectType()) {
+    ty = innerTy;
+    wasOptional = true;
+  }
+  if (!ty->is<GenericTypeParamType>() && !ty->isAnyClassReferenceType())
```

Comment:
I think the check you're doing here ("a class, class existential, class-bound generic or an optional thereof") already exists somewhere as "can I apply ownership to this type". Look around in TypeCheckAttr.cpp when we validate WeakAttr.

Reply:
Yep, thanks!